notes/shelley/Interface.hs

```diff
@@ -0,0 +1,83 @@
+{-# LANGUAGE RecursiveDo #-}
+{-# LANGUAGE Rank2Types #-}
+{-# LANGUAGE BangPatterns #-}
+
+import Control.Monad
+import Control.Concurrent
+import Control.Concurrent.MVar
+import Control.Concurrent.Async
+import Control.Exception
+
+data Logic = Logic
+  { getTip :: IO Int
+  }
+
+data Diffusion = Diffusion
+  { announceBlock :: Int -> IO ()
+  }
+
+data LogicLayer = LogicLayer
+  { logic :: Logic
+  , runLogicLayer :: IO ()
+  }
+
+data DiffusionLayer = DiffusionLayer
+  { diffusion :: Diffusion
+  , runDiffusionLayer :: IO ()
+  }
```

Comment:
Lets document for each of these what it is exactly. For example "the interface provided by the logic layer to the diffusion layer, for the diffusion layer to make calls into the logic layer".